package/metrics_daemon/metrics_daemon/src/metrics_daemon.c

```diff
@@ -0,0 +1,375 @@
+#include <libpiksi/settings.h>
+#include "sbp.h"
+
+#define PROGRAM_NAME "metrics_daemon"
+
+#define METRICS_ROOT_DIRECTORY "/var/log/metrics"
+#define METRICS_OUTPUT_FILENAME "/var/log/metrics.json"
+#define METRICS_USAGE_UPDATE_INTERVAL_MS (10000u)
+
+#define MAX_FOLDER_NAME_LENGTH 64
+static json_object *jobj_root = NULL;
+static json_object *jobj_cur = NULL;
+static char *root_name = NULL;
+static unsigned int root_length = 0;
+static bool first_folder = true;
+static char *target_file = METRICS_OUTPUT_FILENAME;
+const char *metrics_path = METRICS_ROOT_DIRECTORY;
+static bool enable_log_to_file = false;
+
+
+int handle_walk_path(const char *fpath, const struct stat *sb, int tflag);
+char *extract_filename(const char *str);
+json_object *init_json_object(const char *path);
+
+/**
+ * @brief function updates json to file
+ *
+ * @param root       pointer to the root of json object
+ * @param file_name  file to write to
+ */
+static int write_json_to_file(struct json_object *root, const char *file_path)
+{
+  return file_append_string(file_path, json_object_to_json_string(root));
+}
+
+/**
+ * @brief static function that returns file name from path
+ * Ref: Updated code from
+ * https://cboard.cprogramming.com/c-programming/109469-scan-filename-make-list.html
+ *
+ * @param  str        file path in ASCII
+ * @return file_name  file name
+ */
+char *extract_filename(const char *str)
+{
+  int ch = '/';
+  const char *start_ptr = NULL;
+  char *inpfile = NULL;
+  assert(str != NULL);
+  // Search backwards for last backslash in filepath
+  start_ptr = strrchr(str, ch);
+
+  // if backslash not found in filepath
+  if (start_ptr == NULL) {
+    start_ptr = str; // The whole name is a file in current path
+  } else {
+    start_ptr++; // Skip the backslash itself.
+  }
+
+  // extract filename from file path
+  if (start_ptr[0] == '.') start_ptr++;
+  inpfile = calloc(1, MAX_FOLDER_NAME_LENGTH + 1);     // Make space for the zero.
+  strncpy(inpfile, start_ptr, MAX_FOLDER_NAME_LENGTH); // Copy including zero.
+  return inpfile;
+}
+
+
+/**
+ * @brief function that use the full file path loop through the json object tree
+ * to update values
+ *
+ * @param process_path file path in ASCII
+ * @param root         root name, this is used to pair with the root of the json
+ * object tree
+ * @param root_len     number of chars for the root name
+ * @param json_root    root of the json object tree
+ * @return json object node that needs to be updated
+ */
+static struct json_object *loop_through_folder_name(const char *process_path,
+                                                    const char *root,
+                                                    unsigned int root_len)
+{
+  char ch = '/';
+  const char *start_ptr = NULL;
+  bool found_root = false;
+  start_ptr = strchr(process_path, ch); // get the first slash position
+  unsigned int str_len = (unsigned int)strlen(process_path);
+  if (jobj_root == NULL || start_ptr < process_path) {
```

Comment:
I think this should be `start_ptr == NULL` to check if the `/` wasn't found.  Also, I think we can just do `assert(jobj_root != NULL)` since it's always configured in main and it's a programming (or memory) error if it's NULL.